From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvfw32: Implement MCIWNDM_NEW.

MCIWnd windows open a new, empty element of the given device type (or the
current device's) as "open new type ..." does, and notify the new media.
---
 dlls/msvfw32/mciwnd.c | 60 +++++++++++++++++++++++++++++++++++++++++++++++++--
 1 file changed, 58 insertions(+), 2 deletions(-)

diff --git a/dlls/msvfw32/mciwnd.c b/dlls/msvfw32/mciwnd.c
index 7003428..4ff0a7c 100644
--- a/dlls/msvfw32/mciwnd.c
+++ b/dlls/msvfw32/mciwnd.c
@@ -1379,10 +1379,66 @@ end_of_mci_open:
         }
 
     case MCIWNDM_NEWA:
+        {
+            UNICODE_STRING typeW = {0};
+            LRESULT ret;
+
+            TRACE("MCIWNDM_NEWA %s\n", debugstr_a((LPSTR)lParam));
+            if (lParam)
+                RtlCreateUnicodeStringFromAsciiz(&typeW, (LPCSTR)lParam);
+            ret = SendMessageW(hWnd, MCIWNDM_NEWW, wParam, (LPARAM)typeW.Buffer);
+            RtlFreeUnicodeString(&typeW);
+            return ret;
+        }
+
     case MCIWNDM_NEWW:
+        {
+            /* A new, empty element of the given device type (the current
+             * device's when none is given), as "open new type ..." does. */
+            MCI_OPEN_PARMSW mci_open = {0};
+            MCI_GETDEVCAPS_PARMS mci_devcaps;
+            WCHAR type[MAX_PATH], aliasW[64];
+
+            TRACE("MCIWNDM_NEWW %s\n", debugstr_w((LPWSTR)lParam));
+
+            type[0] = 0;
+            if (lParam && *(LPWSTR)lParam)
+                lstrcpynW(type, (LPWSTR)lParam, ARRAY_SIZE(type));
+            else if (mwi->mci)
+                SendMessageW(hWnd, MCIWNDM_GETDEVICEW, ARRAY_SIZE(type), (LPARAM)type);
+            if (!type[0])
+                return MCIERR_MISSING_DEVICE_NAME;
+
+            SendMessageW(hWnd, MCI_CLOSE, 0, 0);
+
+            mci_open.lpstrDeviceType = type;
+            wsprintfW(aliasW, L"%d", HandleToLong(hWnd) + 1);
+            mci_open.lpstrAlias = aliasW;
+            mwi->lasterror = mciSendCommandW(0, MCI_OPEN, MCI_OPEN_TYPE | MCI_OPEN_ALIAS | MCI_WAIT,
+                                             (DWORD_PTR)&mci_open);
+            if (mwi->lasterror)
+            {
+                MCIWND_notify_error(mwi);
+                return mwi->lasterror;
+            }
+
+            mwi->mci = mci_open.wDeviceID;
+            mwi->alias = HandleToLong(hWnd) + 1;
+            mwi->lpName = wcsdup(L"");
+            MCIWND_UpdateState(mwi);
+
+            mci_devcaps.dwItem = MCI_GETDEVCAPS_DEVICE_TYPE;
+            mwi->lasterror = mciSendCommandW(mwi->mci, MCI_GETDEVCAPS, MCI_GETDEVCAPS_ITEM,
+                                             (DWORD_PTR)&mci_devcaps);
+            if (!mwi->lasterror)
+                mwi->dev_type = mci_devcaps.dwReturn;
+            MCIWND_notify_media(mwi);
+            return mwi->lasterror;
+        }
+
     case MCIWNDM_OPENINTERFACE:
-        /* Creating a new element and opening an object interface need MCI
-         * device support that is not available here. */
+        /* Opening an object interface needs MCI device support that is not
+         * available here. */
         FIXME("support for MCIWNDM_ message WM_USER+%d not implemented\n", wMsg - WM_USER);
         return 0;
 
